shared_code/utils.py

```diff
@@ -145,3 +151,50 @@ def locate_potential_duplicate(name, zipcode, client, collection):
     if dupe_candidate is not None:
         return dupe_candidate["name"]
     return False
+
+
+def validate_data(df):
+    # Check if all the required columns ('name', 'address1', 'city', 'state', 'zip', 'serviceSummary')
+    # are in the data frame
+    requiredColumns = ['name', 'address1', 'city', 'state', 'zip', 'serviceSummary']
+    missingColumns = []
+    for column in requiredColumns:
+        if not column in df.columns:
+            missingColumns.append(column)
+    if len(missingColumns) > 0:
+        raise Exception('The data frame does not contain the following required column(s): ' + str(missingColumns))
+
+    columns_to_display = ['name', 'city', 'state', 'zip']
+
+    found_error = False
+
+    # Identifies any row that contains a null or 'NONE' value in the 'name', 'address1', or 'serviceSummary' columns as
+    # invalid
+    for column in ['name', 'address1', 'serviceSummary']:
+        df_errors = df[empty_slots(df, column)]
+        if len(df_errors) > 0:
+            found_error = True
+            logger.error(" null or invalid values found for \'" + column + "\' column in rows: "
+                         + str(df_errors.index.values))
+            logger.error(df_errors[columns_to_display])
+
+    # Identifies any row that contains a null or 'NONE' value in the 'city', 'state', and 'zip' columns as invalid
+    df_errors = df[(empty_slots(df, 'city') | empty_slots(df, 'state')) & empty_slots(df, 'zip')]
+    if len(df_errors) > 0:
+        found_error = True
+        logger.error(" null or invalid values found for \'city\', \'state\', and \'zip\' columns in rows: " +
+                     str(df_errors.index.values))
+        logger.error(df_errors[columns_to_display])
+
+    if found_error:
+        raise Exception('Some rows have invalid data. Check logs for details.')
+
+    # Checks if there are duplicated rows in terms of the columns 'name', 'address1', 'city', 'state', and 'zip'
+    df1 = df[df.groupby(['name', 'address1', 'city', 'state', 'zip'])['serviceSummary'].transform('count') > 1]
+    if len(df1) > 0:
+        logger.error(df1)
+        raise Exception('Duplicate rows have been found. Check logs for details.')
```

Comment:
Let's be specific here in Exception message what we are validating for. We can say , ``There are duplicate entries in data which only differ in `serviceSummary` field and hence can be combined into single entry."

Reply:
Also  what if data really has even the same serviceSummary field i.e. row duplicate.

```
[service1, city1, state1, zip1, summary1] 
[Service1, city1, StaTe1, zip1, SuMMary1] 
```

in this case data is a complete duplicate (note case is Ignored) . Here also we should fail but with message `There are duplicate rows in the data. Check logs for details.`